src/cdcacm.c

```diff
@@ -0,0 +1,364 @@
+}};
+
+static const struct usb_config_descriptor config = {
+    .bLength = USB_DT_CONFIGURATION_SIZE,
+    .bDescriptorType = USB_DT_CONFIGURATION,
+    .wTotalLength = 0,
+    .bNumInterfaces = 3,
+    .bConfigurationValue = 1,
+    .iConfiguration = 0,
+    .bmAttributes = 0xC0,   // self powered
+    .bMaxPower = 50,        // 100mA from USB
+
+    .interface = ifaces,
+};
+
+static const char *usb_strings[] = {
+    "Student Robotics",
+    "Power Board v4",
+    (const char *)SERIALNUM_BOOTLOADER_LOC,
+    "Student Robotics Power Board v4",  // Iface 1
+    "Student Robotics Power Board DFU loader",  // IFace 2, DFU
+};
+
+/* Buffer to be used for control requests. */
+uint8_t usbd_control_buffer[128];
+
+static enum usbd_request_return_codes cdcacm_control_request(usbd_device *usbd_dev, struct usb_setup_data *req, uint8_t **buf,
+        uint16_t *len, void (**complete)(usbd_device *usbd_dev, struct usb_setup_data *req)) {
+    (void)complete;
+    (void)buf;
+    (void)usbd_dev;
+
+    switch(req->bRequest) {
+    case USB_CDC_REQ_SET_CONTROL_LINE_STATE: {
+        /*
+         * This Linux cdc_acm driver requires this to be implemented
+         * even though it's optional in the CDC spec, and we don't
+         * advertise it in the ACM functional descriptor.
+         */
+        char local_buf[10];
+        struct usb_cdc_notification *notif = (void *)local_buf;
+
+        /* We echo signals back to host as notification. */
+        notif->bmRequestType = 0xA1;
+        notif->bNotification = USB_CDC_NOTIFY_SERIAL_STATE;
+        notif->wValue = 0;
+        notif->wIndex = 0;
+        notif->wLength = 2;
+        local_buf[8] = req->wValue & 3;
+        local_buf[9] = 0;
+        // usbd_ep_write_packet(0x83, buf, 10);
+        return USBD_REQ_HANDLED;
+        }
+    case USB_CDC_REQ_SET_LINE_CODING:
+        if(*len < sizeof(struct usb_cdc_line_coding))
+            return USBD_REQ_NOTSUPP;
+
+        return USBD_REQ_HANDLED;
+    case DFU_GETSTATUS:
+        *len = 6;
+        (*buf)[0] = STATE_DFU_IDLE;
+        (*buf)[1] = 100; // ms
+        (*buf)[2] = 0;
+        (*buf)[3] = 0;
+        (*buf)[4] = DFU_STATUS_OK;
+        (*buf)[5] = 0;
+        return USBD_REQ_HANDLED;
+    case DFU_DETACH:
+        re_enter_bootloader = true;
+        return USBD_REQ_HANDLED;
+    }
+    return USBD_REQ_NOTSUPP;
+}
+
+#define USB_MSG_MAXLEN 64
+char usb_msg_buffer[USB_MSG_MAXLEN];
+int usb_msg_len = 0;
+
+static void cdcacm_data_rx_cb(usbd_device *usbd_dev, uint8_t ep) {
+    (void)ep;
+
+    int remaining_len = (USB_MSG_MAXLEN - 1) - usb_msg_len;
+    char* new_data_ptr = usb_msg_buffer + usb_msg_len;
+
+    int len = usbd_ep_read_packet(usbd_dev, 0x01, new_data_ptr, remaining_len);
+
+    if (len) {
+        usb_msg_len += len;
+        usb_msg_buffer[usb_msg_len] = '\0'; // add null terminator to make it a string
+
+        char* end_of_msg = strchr(usb_msg_buffer, '\n');  // test if \n in buffer
+        char response_buffer[64];
```

Comment:
If this needs to be increased in the future, lets ensure it gets done safely.
```suggestion
        #define USB_RESPONSE_MAXLEN 64
        char response_buffer[USB_RESPONSE_MAXLEN];
```